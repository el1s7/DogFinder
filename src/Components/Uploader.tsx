import { Alert, Icon, Input } from 'elementz';
import React from 'react';

import { fileToBase64 } from '../Helpers/Functions';

function Uploader(props: {
	image: HTMLImageElement | string,
	onChange: (base64: string) => void,
	disabled?: boolean
}) {

	const [image, setImage] = [
		props.image instanceof HTMLImageElement ?
			props.image.getAttribute('src') ?? '' :
			props.image,
		props.onChange
	];

	const handleError = () => {
		setImage('');
		Alert.warning('The image file looks invalid');
	};

	const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		if (!e.target.files?.length) {
			return false;
		}

		const file: Blob = e.target.files[0];

		fileToBase64(file).then((base64: ArrayBuffer | string | null) => {
			if (typeof base64 !== 'string') {
				return Promise.reject();
			}

			setImage(base64);
		}).catch(() => (
			Alert.warning("Failed to process image")
		))
	};

	const uploadWrapper = (
		<div className="upload-wrapper upload-overlay">
			<img
				src={image}
				alt="Your dog"
				onError={handleError}
			/>
		</div>
	);

	const uploadEmpty = (
		<div className="upload-empty">
			<Icon name="cloud-upload" />
		</div>
	);

	const uploadBody = image === '' ?
		uploadEmpty :
		uploadWrapper;

	return (
		<Input
			full
			containerClassName="h-100"
			type="file"
			disabled={props.disabled}
			onChange={handleChange}
		>
			{uploadBody}
		</Input>
	);
}

export default Uploader;
