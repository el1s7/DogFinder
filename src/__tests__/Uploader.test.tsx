import React from 'react';
import renderer from 'react-test-renderer';

import Uploader from './../Components/Uploader';

it('renders correctly empty uploader', () => {
	expect.hasAssertions();

	const component = renderer
		.create(
			<Uploader image='' onChange={() => null}/>
		)

	const tree = component.toJSON();

	expect(tree).toMatchSnapshot();
});

it('renders correctly uploaded image', () => {
	expect.hasAssertions();

	const component = renderer
		.create(
			<Uploader image='/akita.jpg' onChange={() => null}/>
		)

	const tree = component.toJSON();

	expect(tree).toMatchSnapshot();
});
