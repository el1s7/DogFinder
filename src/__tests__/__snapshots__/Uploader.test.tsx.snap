// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly empty uploader 1`] = `
<div
  className="ez-input-element input-file h-100 auto"
  onClick={[Function]}
>
  <input
    className="ez-input auto ez-input-wrapped"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="file"
  />
  <div
    className="upload-empty"
  >
    <i
      className="zwicon-cloud-upload"
    />
  </div>
</div>
`;

exports[`renders correctly uploaded image 1`] = `
<div
  className="ez-input-element input-file h-100 auto"
  onClick={[Function]}
>
  <input
    className="ez-input auto ez-input-wrapped"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="file"
  />
  <div
    className="upload-wrapper upload-overlay"
  >
    <img
      alt="Your dog"
      onError={[Function]}
      src="/akita.jpg"
    />
  </div>
</div>
`;
